lib/ansible/plugins/strategy/__init__.py

```diff
@@ -876,14 +876,30 @@ def _do_handler_run(self, handler, handler_name, iterator, play_context, notifie
         host_results = []
         for host in notified_hosts:
             if not iterator.is_failed(host) or play_context.force_handlers:
-                task_vars = self._variable_manager.get_vars(play=iterator._play, host=host, task=handler)
-                self.add_tqm_variables(task_vars, play=iterator._play)
-                self._queue_task(host, handler, task_vars, play_context)
+                if handler.action == 'meta':
+                    meta_name = handler.args.get('_raw_params', None)
+                    # FIXME filter others?
+                    if meta_name in ('flush_handlers'):
+                        display.error('%s not supported as a handler, skipping...' % meta_name)
+                        return_data  = {'msg': '%s not supported as a handler' % meta_name, 'skipped': True}
+                        host_result = TaskResult(host, handler, return_data)
+                        host_results.append(host_result)
+                        continue
+
+                    if meta_name not in ('noop', 'reset_connection', 'end_host'):
+                        run_once = True
```

Comment:
should we allow 'run once' tasks? since they affect hosts globally it is kind of counter to 'notify for specific hosts'

Reply:
We do allow `run_once` on 'normal' handler tasks though so I am not sure. I can't come up with use cases for the `run_once` meta tasks in a handler...